In [1]:
import gensim
import numpy as np
import pandas as pd
import os
import re
from nltk.tokenize import TreebankWordTokenizer

pd.set_option('display.max_colwidth', None)

In [2]:
shake_bi = pd.read_csv('../data/shakespearean_bigrams_pos.csv')

In [3]:
shake_bi.sample(5)

title               title_tokenized        pos_tag  \
205373              thou note              ['thou', 'note']   ['NN', 'NN']   
192052           the blackest           ['the', 'blackest']  ['DT', 'JJS']   
114350              look they              ['look', 'they']  ['NN', 'PRP']   
182886  strange insurrections  ['strange', 'insurrections']  ['JJ', 'NNS']   
152805               pluck 'd               ['pluck', "'d"]   ['NN', 'MD']   

                  pos  pos_len pos1 pos2  
205373   ['NN', 'NN']        2   NN   NN  
192052  ['DT', 'JJS']        2   DT  JJS  
114350  ['NN', 'PRP']        2   NN  PRP  
182886  ['JJ', 'NNS']        2   JJ  NNS  
152805   ['NN', 'MD']        2   NN   MD

In [4]:
shake_bi = shake_bi.drop(columns=['title_tokenized', 'pos_tag'])

In [5]:
twd = TreebankWordTokenizer()
shake_bi['title_token'] = shake_bi['title'].apply(lambda x: twd.tokenize(x))
shake_bi.sample(5)

title             pos  pos_len  pos1 pos2       title_token
198858   their tender  ['PRP$', 'NN']        2  PRP$   NN   [their, tender]
101585  instigated by   ['VBN', 'IN']        2   VBN   IN  [instigated, by]
30054     bought mine    ['NN', 'NN']        2    NN   NN    [bought, mine]
18986         as they   ['IN', 'PRP']        2    IN  PRP        [as, they]
144203        or from    ['CC', 'IN']        2    CC   IN        [or, from]

Previous work has demonstrated that words with a period at the end (such as `soldier.`) will not have an accompanying word vector, which would result in the loss of thousands of entries. The below code will remove the period at the end of such words, but will retain other punctuation and single event periods.

In [16]:
titles = shake_bi.title_token.tolist()

In [7]:
first = []
second = []

for title in titles:
    first.append(title[0])
    second.append(title[1])
        

In [9]:
new_first = []

for word in first:
    if len(word) > 1 and word.endswith('.'):
        new_first.append(word.rstrip(word[-1]))
    else:
        new_first.append(word)
        

In [10]:
new_second = []

for word in second:
    if len(word) > 1 and word.endswith('.'):
        new_second.append(word.rstrip(word[-1]))
    else:
        new_second.append(word)

In [11]:
title_clean = list(list(x) for x in zip(new_first, new_second))

In [14]:
title_clean[177240:177245]

[['soldier', 'with'],
 ['soldier', 'you'],
 ['soldier-like', 'phrase'],
 ['soldier', 'and'],
 ['soldier', 'let']]

In [15]:
shake_bi['title_clean'] = title_clean
shake_bi.sample(5)

title           pos  pos_len pos1 pos2     title_token  \
210605   to example  ['TO', 'NN']        2   TO   NN   [to, example]   
82764        hate '  ['NN', '""']        2   NN   ""       [hate, ']   
93104       hit the  ['VB', 'DT']        2   VB   DT      [hit, the]   
61280   every coast  ['DT', 'NN']        2   DT   NN  [every, coast]   
105881     jove and  ['NN', 'CC']        2   NN   CC     [jove, and]   

           title_clean  
210605   [to, example]  
82764        [hate, ']  
93104       [hit, the]  
61280   [every, coast]  
105881     [jove, and]

Now to replace every word in each bigram with the learned word vector

In [18]:
wv = gensim.models.KeyedVectors.load("shake_w2v.wordsvectors", mmap='r')

In [19]:
words = set(wv.index2word)
len(words)

26304

In [20]:
bigram_vect = [np.array([wv[i] for i in ls if i in words])
                                for ls in shake_bi['title_clean']]

In [23]:
loss_title_clean = []

for i,v in enumerate(bigram_vect):
    if len(v) != 2:
        loss_title_clean.append(i)
#         print(shake_bi['title_clean'][i])
#         print('Vector len:', len(v), 'Index:', i)

print('Number of bigrams with no word matches form title_clean', len(loss_title_clean))

Number of bigrams with no word matches form title_clean 67


In [24]:
bigram_vect2 = [np.array([wv[i] for i in ls if i in words])
                                for ls in shake_bi['title_token']]

In [25]:
loss_title_token = []

for i,v in enumerate(bigram_vect2):
    if len(v) != 2:
        loss_title_token.append(i)

print('Number of bigrams with no word matches form title_token', len(loss_title_token))

Number of bigrams with no word matches form title_token 5970


After all the above coding, I think I would rather lose 5,970 titles. Now that it is morning, and I am thinking *clearly*, I would not want most of those titles anyway because they are at the end of a sentence. I think it was my original belief, that genism took sentence ends into consideration. Reversing course below...

In [29]:
shake_bi.drop(columns=['title_clean'], inplace=True)
shake_bi.sample()

title            pos  pos_len pos1 pos2      title_token
120676  me something  ['PRP', 'NN']        2  PRP   NN  [me, something]

In [30]:
shake_bi['word_vec'] = bigram_vect2
shake_bi.sample()

title            pos  pos_len pos1 pos2 title_token  \
237925  ye done  ['NN', 'VBN']        2   NN  VBN  [ye, done]   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      word_vec  
237925  [[0.38721955, -0.6298508, -0.64152753, 0.5480759, -0.27110818, -0.39844513, -0.043123763, -0.7947163, 0.72984946, -0.16564111, -0.33992615, -0.92562103, 0.14775954, -0.36516213, -0.5071994, 0.099314995, -0.22760622, 0.3127083, 0.07279155, -0.5925049, 0.19713917, -0.09201206, 0.11940089, -0.17175232, 0.54297876, 0.31815404, -0.19652726, 0.5672226, 0.78805965, 0.24863446, 0.85366786, 0.31839156, -0.29994288, 0.48651826, -0.16115738, 0.30141068, 0.8820832, -0.791511, -0.13478667, -0.34277233, 0.43665951, -0.39968136, 0.1799807, -0.36762482, 0.3285105, -0.0061254804, -0.088755235, 0.29515928, -0.016218198, 0.013438691, 0.13234694, 0.072043456, -0.117082514, 0.18616527, -0.49476078, -0.15034261, 0.34449473, -0.7079535, -0.0849729, 0.7924161, 0.53180283, 0.022599643, -0.73220396, -0.3911725, 0.48320225, 0.20673294, 0.7856577, 0.011493572, 0.43976203, -0.32445842, 0.08919123, -0.6732706, 0.003777869, 0.56608033, 0.9314576, -0.13060005, -0.088135466, -0.108308144, 0.27884117, -0.46055293, 0.07342696, 0.0559225, 0.69816613, 0.4011239, -0.028544694, 0.08064641, 0.026036564, -0.19206068, -0.41460633, -0.4037335, 0.2626454, -0.12135888, 0.3745098, -0.13898301, 0.054465998, -0.116109446, 0.11705433, 0.4324808, -0.12951922, 0.22345495, ...], [-0.07660817, -0.46201295, -0.5043615, -0.37414637, -0.31470814, -0.8197248, -0.44553378, -0.5021222, -0.2418622, -0.031998783

In [31]:
shake_bi.shape

(242601, 7)

In [32]:
len(loss_title_token)

5970

In [33]:
shake_bi.drop(index=loss_title_token, inplace=True)
shake_bi.shape

(236631, 7)

In [34]:
shake_bi.reset_index(drop=True, inplace=True)

In [35]:
datapath = '../data'

datapath_shake_bi = os.path.join(datapath, 'shakespearean_bigrams_pos_w2v.csv')
shake_bi.to_csv(datapath_shake_bi, index=False)